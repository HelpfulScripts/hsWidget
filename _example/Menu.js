import { RadioButton } from './RadioButton';
export class Menu extends RadioButton {
    view(node) { return this.viewGroup('.hs-menu', node); }
}
;
//# sourceMappingURL=data:application/json;base64,eyJ2ZXJzaW9uIjozLCJmaWxlIjoiTWVudS5qcyIsInNvdXJjZVJvb3QiOiIiLCJzb3VyY2VzIjpbIi4uL3NyYy9NZW51LnRzIl0sIm5hbWVzIjpbXSwibWFwcGluZ3MiOiJBQXdEQSxPQUFPLEVBQUUsV0FBVyxFQUFFLE1BQU8sZUFBZSxDQUFDO0FBTTdDLE1BQU0sV0FBWSxTQUFRLFdBQVc7SUFDakMsSUFBSSxDQUFDLElBQVcsSUFBVyxPQUFPLElBQUksQ0FBQyxTQUFTLENBQUMsVUFBVSxFQUFFLElBQUksQ0FBQyxDQUFDLENBQUMsQ0FBQztDQUN4RTtBQUFBLENBQUMifQ==