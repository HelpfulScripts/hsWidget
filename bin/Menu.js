"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const RadioButton_1 = require("./RadioButton");
class Menu extends RadioButton_1.RadioButton {
    view(node) { return this.viewGroup('.hs-menu', node); }
}
exports.Menu = Menu;
;
//# sourceMappingURL=data:application/json;base64,eyJ2ZXJzaW9uIjozLCJmaWxlIjoiTWVudS5qcyIsInNvdXJjZVJvb3QiOiIiLCJzb3VyY2VzIjpbIi4uL3NyYy9NZW51LnRzIl0sIm5hbWVzIjpbXSwibWFwcGluZ3MiOiI7O0FBd0RBLCtDQUE2QztBQU03QyxVQUFrQixTQUFRLHlCQUFXO0lBQ2pDLElBQUksQ0FBQyxJQUFXLElBQVcsT0FBTyxJQUFJLENBQUMsU0FBUyxDQUFDLFVBQVUsRUFBRSxJQUFJLENBQUMsQ0FBQyxDQUFDLENBQUM7Q0FDeEU7QUFGRCxvQkFFQztBQUFBLENBQUMifQ==