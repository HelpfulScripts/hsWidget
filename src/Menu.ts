/**
 * # Menu Widget
 * Creates a simple menu with several items.
 * 
 * ### Profile
 * invoked as `m(Menu, { desc:<MenuDesc> })`
 * 
 * ### Attributes (node.attrs):
 * - `desc:` {@link Menu.MenuDesc MenuDesc}
 *     - `items: string[]`                  the items on the menu
 *     - `changed: (item:string) => void`   called when item clicked
 *     - `defaultItem?: number|string`      the currently selected item, by index or name
 *     - `itemCSS?: string[]`               css to apply to items;
 * - `css?: string`                         css class to assign to button group
 * - `style?: string`                       style string to apply to button tag
 * - `size?: string | string[]`             sizes to layout menu items; 
 * 
 * ### Example
 * <example>
 * <file name='script.js'>
 * const items = ['One', 'Two', 'Three'];
 * const content   = ['1st', '2nd', '3rd'];
 * let  theContent = content[1];
 * 
 * m.mount(root, {view: () => m(hslayout.Layout, {
 *     rows:["30px", "fill"],
 *     content:[
 *         m(hswidget.Menu, {desc: {
 *             items: items,
 *             defaultItem: 'Two',
 *             changed: item => 
 *                theContent = content[items.indexOf(item)]
 *         }}),
 *         m(hslayout.Layout, { css:'myMain', content: theContent })
 *     ]
 * })});
 *
 * </file>
 * <file name='style.css'>
 * .myMain { 
 *    border:1px solid #ddd;
 *    border-top: 0px solid #ddd;
 * } 
 * .hs-selectable { 
 *     background-color: #f4f4e8; 
 * }
 * .hs-selected { 
 *     background-color: #eed; 
 *     border-width:0px;
 * }
 * </file>
 * </example>
 */

 /** */
import { Vnode }        from 'hslayout';
import { RadioButton }  from './RadioButton';


/**
 * Creates a simple menu with several items, as configured by the desc:SelectorDesc object passed as a parameter. 
 */
export class Menu extends RadioButton {
    view(node: Vnode): Vnode { return this.viewGroup('.hs-menu', node); }
};
