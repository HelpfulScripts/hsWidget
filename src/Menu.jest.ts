import { Menu, MenuPanel } from './Menu';
import m from "mithril";

const left  = ['0%', '25%', '50%', '75%'];
const right = ['75%', '50%', '25%', '0%'];
const items = ['1a', '2a', '3a', '4a'];

window = Object.assign(require('mithril/test-utils/domMock.js')(), require('mithril/test-utils/pushStateMock')());
const mq = require('mithril-query');

describe('Menu', () => {
    const out = mq(m(Menu, {
        onclick: (item:number) => { console.log('selected'); }
    }, items));

    test ('DOM structure', () => {
        out.should.have('.hs_menu');
        out.should.have(4, '.hs_menu>.hs_button');
        out.should.have(3, '.hs_menu>.hs_button.state0');    
        out.should.have(1, '.hs_menu>.hs_button.state1');    
    });
});

const root = window.document.createElement("div");

describe('MenuPanel', () => {
    beforeEach(() =>
        m.mount(root, {view: () => m(MenuPanel, {
            menu: ["one", "two", "three"],  
            initial: 1
        }, ['1st', '2nd', '3rd'])})
    );

    test('layout', () => expect(root).toMatchSnapshot());
});