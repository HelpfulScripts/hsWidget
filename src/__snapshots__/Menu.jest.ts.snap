// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MenuPanel layout 1`] = `
<div>
  <div
    class="hs_grid  hs_grid_row"
    style="grid-template-rows: 30px auto;"
  >
    <div
      class="hs_option_buttons  hs_radio_buttons  hs_menu"
      style="grid-template-columns: repeat(3, 1fr);"
    >
      <div
        class="hs_button state0  hs_onoff_button index0"
      >
        one
      </div>
      <div
        class="hs_button state1  hs_onoff_button index1"
      >
        two
      </div>
      <div
        class="hs_button state0  hs_onoff_button index2"
      >
        three
      </div>
    </div>
    <div
      class="hs_grid_cell child1"
    >
      1st
    </div>
  </div>
</div>
`;
